### PR TITLE
Simplify / cleanup async demo

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR simplifies the async demo by using a CancellationTokenSource.

It also reformats the code and updates the method names to be more standard C#.

Tested with Visual Studio 2022 + .NET Framework 4.8.

### CODE REVIEW

File: csharp/Ice/async/Client.cs [hunk 1]
```diff
@@ -15,116 +14,107 @@ public static int Main(string[] args)
...
-                    menu();
+                    case null:
+                        break;
+
+                    case "i":
+                        hello.sayHello(0);
+                        break;
+
+                    case "d":
+                        _ = HelloAsync(hello);
```
Comment (pepone) [inline csharp/Ice/async/Client.cs:68]: This doesn't work with Visual Studio 2015, .NET 4.5

We can do `Task.Run(() => HelloAsync(hello));`

File: csharp/Ice/async/WorkQueue.cs
Comment (pepone) [inline csharp/Ice/async/WorkQueue.cs:1]: The project still references this removed file

File: csharp/Ice/async/Server.cs [hunk 2]
```diff
@@ -3,60 +3,55 @@
...
+                        };
 
-                    var adapter = communicator.createObjectAdapter("Hello");
-                    adapter.add(new HelloI(workQueue), Ice.Util.stringToIdentity("hello"));
-
-                    workQueue.Start();
-                    try
-                    {
+                        Ice.ObjectAdapter adapter = communicator.createObjectAdapter("Hello");
+                        _ = adapter.add(new HelloI(cts), Ice.Util.stringToIdentity("hello"));
```
Comment (pepone) [inline csharp/Ice/async/Server.cs:35]: `_` doesn't work with VS 2015

File: csharp/Ice/async/Server.cs [hunk 3]
```diff
@@ -3,60 +3,55 @@
 //
 
 using System;
+using System.Threading;
+using System.Threading.Tasks;
 
 public class Server
 {
-    public static int Main(string[] args)
+    public static async Task<int> Main(string[] args)
```
Comment (pepone) [inline csharp/Ice/async/Server.cs:11]: Visual Studio 2015 doesn't allow async main

File: csharp/Ice/async/Client.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/Ice/async/Client.cs]: Fixed, but didn't use Task.Run.

File: csharp/Ice/async/WorkQueue.cs
Comment (bernardnormier) [reply csharp/Ice/async/WorkQueue.cs:1]: Fixed.

File: csharp/Ice/async/Server.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/Ice/async/Server.cs]: Removed.

File: csharp/Ice/async/Server.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/Ice/async/Server.cs]: Fixed and tested with VS 2015.

**Comment by bernardnormier** [comment]
Tested with VS 2015 Update 3.

**Review by bentoi — APPROVED** [review]

File: csharp/Ice/async/Client.cs [hunk 4]
```diff
@@ -15,121 +14,107 @@ public static int Main(string[] args)
 
         try
         {
-            //
-            // using statement - communicator is automatically destroyed
-            // at the end of this statement
-            //
-            using(var communicator = Ice.Util.initialize(ref args, "config.client"))
+            using (Ice.Communicator communicator = Ice.Util.initialize(ref args, "config.client"))
```
Comment (externl) [inline csharp/Ice/async/Client.cs:17]: Why get rid of the comment?

File: csharp/Ice/async/README.md [hunk 5]
```diff
@@ -19,12 +19,9 @@ client
 > dotnet client.dll
 > ```
 
-The demo invocation can either have a short response time or require a
-significant amount of time to complete. For the long running request
-the client uses AMI and the server uses AMD plus a worker thread to
-process the request. While a long request is processing, short
-requests are still able to be processed and more long requests can be
-queued for processing by the worker thread.
+The shutdown of an server (and its object adapter) waits for outstanding dispatches
```
Comment (externl) [inline csharp/Ice/async/README.md:22]:
```suggestion
The shutdown of a server (and its object adapter) waits for outstanding dispatches
```

File: csharp/Ice/async/Server.cs [hunk 6]
```diff
@@ -12,51 +14,38 @@ public static int Main(string[] args)
...
-                        communicator.waitForShutdown();
-                    }
-                    finally
-                    {
-                        workQueue.Join();
+
+                        // cts is canceled by Ctrl+C or a shutdown request.
+                        // In modern C#, you should call: await Task.Delay(-1, cts.Token)
+                        cts.Token.WaitHandle.WaitOne();
+                        communicator.shutdown();
```
Comment (externl) [inline csharp/Ice/async/Server.cs:41]: I would not say "in modern C#". Or at least say that we need to do it this way for compatibility with older versions of .NET

**Review by externl — APPROVED** [review]

File: csharp/Ice/async/Client.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/Ice/async/Client.cs:17]: What `using` is about is basic C#. It's odd to have this comment in every single server.

File: csharp/Ice/async/Server.cs (hunk 6 above)
Comment (bernardnormier) [reply csharp/Ice/async/Server.cs:41]: Fixed

**Pull request merged**